### DISCUSSION TITLE
No inlay type hints when opening a file from the command line

### DISCUSSION THREAD

**Discussion by cfmantilla** [helix-editor/helix#9585] [Troubleshooting]
If I open a file from the command line: 
hx src/main.rs

Checking the log file it shows this error:
2024-02-05T10:10:22.508 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(153)), original request likely timed out

If I open the file using the file picker it works.

This is part of the log:
 2024-02-02T17:51:39.254 helix_term::application [WARN] Ignoring a client/registerCapability request because dynamic capability registration is not enabled. Please report this upstream
       │  to the language server
   2   │ 2024-02-05T10:10:22.508 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(153)), original request likely timed out
   3   │ 2024-02-05T10:25:22.962 helix_lsp::transport [ERROR] rust-analyzer err <- "Panic context:\n"

Platform: Linux POP_OS

Terminal emulator: Kitty

Version: 23.10 (d570c29c), installed by cloning the git repo

**Comment by the-mikedavis** [comment]
What does the log say when running in verbose mode? `hx -v --log log.txt` and attach `log.txt`.

Please use the bug report template in the future.

**Reply by cfmantilla** [reply]
Sure, this is the log file:
[log.txt](https://github.com/helix-editor/helix/files/14226876/log.txt)

**Comment by cfmantilla** [comment]
this is another log file
first I open lib.rs:  hx -v --log log.txt src/lib.rs

I can't see type hints:

![image](https://github.com/helix-editor/helix/assets/7266440/68b3192d-9c09-47db-a93e-8deea81890ef)

then I open main.rs using the file picker and it shows inlays:
![image](https://github.com/helix-editor/helix/assets/7266440/af3574a6-e124-4134-83fc-728caf48778e)

then I close the lib.rs buffer and reopen it  using the file picker, it shows type hints:
![image](https://github.com/helix-editor/helix/assets/7266440/10567757-476c-4b21-834b-dff5a0a5be7e)

then I quit helix, this is the log file:
[log.txt](https://github.com/helix-editor/helix/files/14273935/log.txt)

**Comment by the-mikedavis** [comment]
I can reproduce this. The log has the lines:

```
2024-02-13T19:19:20.723 helix_lsp::transport [INFO] rust-analyzer -> {"jsonrpc":"2.0","method":"textDocument/inlayHint","params":{"range":{"end":{"character":0,"line":99},"start":{"character":0,"line":0}},"textDocument":{"uri":"file:///home/carlosmantilla/workspace/rust/multi_threaded_http_server/src/lib.rs"}},"id":1}
2024-02-13T19:19:20.724 helix_lsp::transport [INFO] rust-analyzer <- {"jsonrpc":"2.0","id":1,"result":[]}
```

We are requesting inlay hints and rust-analyzer is saying that the document doesn't have any.

Starting Helix with the file argument isn't the only way to reproduce this: it happens when we request inlay hints from rust-analyzer before it's fully ready. (See the progress messages in the log file - we're requesting inlay hints before rust-analyzer is done indexing the project.) For example starting `hx`, opening the file picker and then choosing `src/lib.rs` has the same behavior. You can kick off another request for inlay hints by editing the file or by closing it and reopening as you mentioned.

Without knowing the internals of rust-analyzer it seems to me that rust-analyzer should wait until it's done starting up and indexing to respond to inlay hint requests. I'm not sure there's a robust way to fix this on the Helix side other than maybe re-requesting inlay hints after we get a publishDiagnostic message from the server. @pascalkuthe do you have an opinion on this? Should this be fixed upstream?

**Reply by pascalkuthe** [reply]
This may be a rust-analyzer bug but mayberewriting inlay hints to use the event system fixes this. With that we could retry the inlay hints request if it fails (with somekind of exponentinal backoff). We should figure out if/how vscode Handels that. If the same bug exists there it's an upstream issue if not better async Handels would help
